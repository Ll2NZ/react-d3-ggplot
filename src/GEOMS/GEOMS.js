import React            from "react";
import PropTypes        from "prop-types";
import { ChartContext } from "../_context/";

function GEOMS(props) {
        return(
            <ChartContext.Provider value={ props }>
                <svg width={ props.dimensions.width } height={ props.dimensions.height }>
                    { props.children }
                </svg>
            </ChartContext.Provider>
        );
}

GEOMS.propTypes = {
    dimensions: PropTypes.object,
    children: PropTypes.any
};

export default GEOMS;
