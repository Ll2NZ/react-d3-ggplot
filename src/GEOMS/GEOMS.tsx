import React, { ReactChildren, ReactComponentElement, ReactChild } from "react";
import { ChartContext, IContext } from "../_context/";

interface IProps extends IContext {
  children?: ReactChild[] | ReactChild;
}

function GEOMS(props: IProps) {
  if (!props.data || !props.aes || !props.dimensions) {
    throw new Error(`GEOMS must have the following properties: data, aes, and dimension`);
  }

  if (!props.children) {
    throw new Error("GEOMS expects at least one child components such as: <Rects/> or <Points/>");
  }

  return (
    <ChartContext.Provider value={props}>
      <svg width={props.dimensions.width} height={props.dimensions.height}>
        {props.children}
      </svg>
    </ChartContext.Provider>
  );
}

export default GEOMS;
