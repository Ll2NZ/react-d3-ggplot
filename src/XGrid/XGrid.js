import React            from "react";
import { useContext }   from "react";
import { useRef }       from "react";
import { useEffect }    from "react";
import { useScale }     from "../_hooks/";
import { ChartContext } from "../_context/";
import { select }       from "d3-selection";
import { axisBottom }   from "d3-axis";

function XGrid(props) {
    const node    = useRef(null);
    const context = useContext(ChartContext);
    const scale   = useScale(context, XGrid.name);

    useEffect(() => {
        // find grid line locations
        const { dimensions } = context;
        const axisPosition   = `translate(0, ${(dimensions.height - dimensions.padding)})`;

        // append XGrid lines
        select(node.current)
            .attr("transform", axisPosition)
            .call(
                axisBottom(scale)
                .ticks(4)
                .tickSize(-(dimensions.height - dimensions.padding*2))
                .tickFormat("")
            );

        // color grid lines
        select(node.current)
            .selectAll("g")
            .selectAll("line")
            .attr("stroke", "rgb(255, 255, 255)");
    });

    return(
        <g ref={ node }>
        </g>
    );
}

export default XGrid;
