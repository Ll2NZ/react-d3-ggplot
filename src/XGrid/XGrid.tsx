import React from "react";
import { useContext } from "react";
import { useRef } from "react";
import { useEffect } from "react";
import { useXScale } from "../_hooks/";
import { ChartContext } from "../_context/";
import { select } from "d3-selection";
import { axisBottom } from "d3-axis";

interface IProps {
  fill?: string;
}

XGrid.displayName = "XGrid";

function XGrid(props: IProps) {
  const ref: any = useRef(null);
  const { dimensions } = useContext(ChartContext);
  const xScale: any = useXScale();
  const tickPosition = `translate(0, ${dimensions.padding})`;

  useEffect(() => {
    select(ref.current)
      .attr("transform", tickPosition)
      .call(
        axisBottom(xScale)
          .ticks(4)
          .tickSize(dimensions.height - dimensions.padding * 2)
          .tickSizeOuter(0)
      );

    select(ref.current)
      .select("path")
      .remove();

    select(ref.current)
      .selectAll("text")
      .remove();

    select(ref.current)
      .selectAll("line")
      .attr("stroke", "white")
      .attr("opacity", 0.7);
  });

  return <g ref={ref} />;
}

export default XGrid;
