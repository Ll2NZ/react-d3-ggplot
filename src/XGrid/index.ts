export { default as XGrid } from "./XGrid";
