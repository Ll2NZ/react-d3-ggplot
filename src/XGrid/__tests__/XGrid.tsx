import * as React from "react";
import { render, cleanup } from "react-testing-library";
import { GEOMS } from "../../GEOMS/";
import XGrid from "../XGrid";

afterEach(cleanup);

const validState = {
  data: [{ x: 0, y: 0 }, { x: 10, y: 10 }],
  aes: ["x", "y"],
  dimensions: { width: 500, height: 400, padding: 50 }
};

test("Should render multiple <line> elements", () => {
  const node = render(
    <GEOMS {...validState}>
      <XGrid />
    </GEOMS>
  );
});

test("Underlyng markup should not change as we change our logic in code base unless we chage the markup ourselves", () => {
  const { container } = render(
    <GEOMS {...validState}>
      <XGrid />
    </GEOMS>
  );

  expect(container.firstChild).toMatchSnapshot();
});
