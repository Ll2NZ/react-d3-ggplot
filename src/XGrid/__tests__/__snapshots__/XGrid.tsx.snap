// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Underlyng markup should not change as we change our logic in code base unless we chage the markup ourselves 1`] = `
<svg
  height="400"
  width="500"
>
  <g
    fill="none"
    font-family="sans-serif"
    font-size="10"
    text-anchor="middle"
    transform="translate(0, 50)"
  >
    <g
      class="tick"
      opacity="1"
      transform="translate(50.5,0)"
    >
      <line
        opacity="0.7"
        stroke="white"
        y2="300"
      />
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(130.5,0)"
    >
      <line
        opacity="0.7"
        stroke="white"
        y2="300"
      />
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(210.5,0)"
    >
      <line
        opacity="0.7"
        stroke="white"
        y2="300"
      />
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(290.5,0)"
    >
      <line
        opacity="0.7"
        stroke="white"
        y2="300"
      />
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(370.5,0)"
    >
      <line
        opacity="0.7"
        stroke="white"
        y2="300"
      />
    </g>
    <g
      class="tick"
      opacity="1"
      transform="translate(450.5,0)"
    >
      <line
        opacity="0.7"
        stroke="white"
        y2="300"
      />
    </g>
  </g>
</svg>
`;
